crates/sui-bridge/src/eth_syncer.rs

```diff
@@ -0,0 +1,346 @@
+            let last_finalized_block_rx_clone = last_finalized_block_rx.clone();
+            let eth_client_clone = self.eth_client.clone();
+            task_handles.push(spawn_logged_monitored_task!(
+                Self::run_event_listening_task(
+                    contract_address,
+                    start_block,
+                    last_finalized_block_rx_clone,
+                    eth_evnets_tx_clone,
+                    eth_client_clone,
+                )
+            ));
+        }
+        Ok((task_handles, eth_events_rx, last_finalized_block_rx))
+    }
+
+    async fn run_finalized_block_refresh_task(
+        last_finalized_block_sender: watch::Sender<u64>,
+        eth_client: Arc<EthClient<P>>,
+    ) {
+        tracing::info!("Starting finalized block refresh task.");
+        let mut last_block_number = 0;
+        let mut interval = time::interval(FINALIZED_BLOCK_QUERY_INTERVAL);
+        interval.set_missed_tick_behavior(time::MissedTickBehavior::Skip);
+        loop {
+            interval.tick().await;
+            let new_value = retry_with_max_delay!(
+                eth_client.get_last_finalized_block_id(),
+                Duration::from_secs(600)
+            )
+            .expect("Failed to get last finalzied block from eth client after retry");
+            tracing::debug!("Last finalized block: {}", new_value);
+            if new_value > last_block_number {
+                last_finalized_block_sender
+                    .send(new_value)
+                    .expect("last_finalized_block channel receiver is closed");
+                tracing::info!("Observed new finalized eth block: {}", new_value);
+                last_block_number = new_value;
+            }
+        }
+    }
+
+    async fn run_event_listening_task(
+        contract_address: ethers::types::Address,
+        mut start_block: u64,
+        mut last_finalized_block_receiver: watch::Receiver<u64>,
+        events_sender: mysten_metrics::metered_channel::Sender<Vec<ethers::types::Log>>,
+        eth_client: Arc<EthClient<P>>,
+    ) {
+        tracing::info!(contract_address=?contract_address, "Starting eth events listening task from block {start_block}");
+        loop {
+            last_finalized_block_receiver
+                .changed()
+                .await
+                .expect("last_finalized_block channel sender is closed");
+            let new_finalized_block = *last_finalized_block_receiver.borrow();
+            if new_finalized_block < start_block {
+                tracing::info!(
+                    contract_address=?contract_address,
+                    "New finalized block {} is smaller than start block {}, ignore",
+                    new_finalized_block,
+                    start_block,
+                );
+                continue;
+            }
+            let events = retry_with_max_delay!(
+                eth_client.get_events_in_range(contract_address, start_block, new_finalized_block),
```

Comment:
How many events are there typically per block, and how many blocks per second?

Reply:
this function only queries events in the given contract_address, so i'd expect lower than 1 per second, if that's what you were asking